### PR TITLE
Lookup in extended type (simple name and member access)

### PR DESCRIPTION
**Pull request by jcouv**
This PR updates member lookup on extensions to include members from the underlying type. The corresponding section of the spec:

> ### Base types
> 
> TL;DR: Member lookup on an extension type includes members from its base extensions, its extended type and base types.
> 
> We modify the [base types rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1252-base-types) as follows:
> 
> For purposes of member lookup, a type `T` is considered to have the following base types:
> 
> - If `T` is `object` or `dynamic`, then `T` has no base type.
> - If `T` is an *enum_type*, the base types of `T` are the class types `System.Enum`, `System.ValueType`, and `object`.
> - If `T` is a *struct_type*, the base types of `T` are the class types `System.ValueType` and `object`.
> - If `T` is a *class_type*, the base types of `T` are the base classes of `T`, including the class type `object`.
> - If `T` is an *interface_type*, the base types of `T` are the base interfaces of `T` and the class type `object`.
> - If `T` is an *array_type*, the base types of `T` are the class types `System.Array` and `object`.
> - If `T` is a *delegate_type*, the base types of `T` are the class types `System.Delegate` and `object`.
> - \***If `T` is an *extension_type*, the base types of `T` are the base extensions of `T` and the extended type of `T` and its base types.**

~~Question: I'm debating whether to go ahead and update `BaseTypeNoUseSiteDiagnostics` and `GetNextBaseTypeNoUseSiteDiagnostics` already...~~ From offline discussion, we won't want to update those APIs since there is no inheritance relationship between an extension and its base extensions or its underlying type.

Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

**Comment by jcouv** [comment]
@AlekseyTs @jjonescz for review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -7788,6 +7794,7 @@ private static void CombineExtensionMethodArguments(BoundExpression receiver, An
             foreach (var scope in new ExtensionScopes(this))
             {
                 if (!left.Type.IsTypeParameter()
+                    && !left.Type.IsExtension
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7797]: >&& !left.Type.IsExtension [](http://example.com/codeflow?start=20&length=25)

Could you please elaborate? Why this change is needed? Why it is not needed in the next `if`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -7788,6 +7794,7 @@ private static void CombineExtensionMethodArguments(BoundExpression receiver, An
             foreach (var scope in new ExtensionScopes(this))
             {
                 if (!left.Type.IsTypeParameter()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7796]: >if (!left.Type.IsTypeParameter() [](http://example.com/codeflow?start=16&length=32)

Could you please add comments. Why there are two somewhat similar, but different `if` statements in this loop? I am reading the code and realizing that I have no idea why things are done in this way.  Why type checks are important in the first `if`, but not in the second, etc. It would be good to clear capture the intent and explain how implementation follows/matches it.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 3]
```diff
@@ -1449,6 +1463,29 @@ private void MergeHidingLookupResults(LookupResult resultHiding, LookupResult re
             {
                 resultHiding.MergePrioritized(resultHidden);
             }
+
+            return;
+
+            static bool isInExtensionHierarchy(NamedTypeSymbol hiddenContainer, NamedTypeSymbol hidingContainer, ConsList<TypeSymbol> basesBeingResolved, CSharpCompilation compilation, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1469]: >isInExtensionHierarchy [](http://example.com/codeflow?start=24&length=22)

Could you please add a comment explaining expected behavior of the method, i.e. stating the intent?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 4]
```diff
@@ -1449,6 +1463,29 @@ private void MergeHidingLookupResults(LookupResult resultHiding, LookupResult re
...
+            {
+                if (!hidingContainer.IsExtension)
+                {
+                    return false;
+                }
+
+                var extendedType = hidingContainer.ExtendedTypeNoUseSiteDiagnostics;
+                if (hiddenContainer.Equals(extendedType, TypeCompareKind.ConsiderEverything))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1477]: >TypeCompareKind.ConsiderEverything [](http://example.com/codeflow?start=57&length=34)

Why not ignore everything? Do we have a test demonstrating significance of this option?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1469]: Perhaps the intent would be obvious if the helper was named something like "isOrDerivedFromExtendedType".

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs [hunk 5]
```diff
@@ -230,7 +230,6 @@ private ExtensionInfo MakeDeclaredExtensionInfo(ConsList<TypeSymbol>? basesBeing
             TypeSymbol? underlyingType = null;
             SourceLocation? underlyingTypeLocation = null;
 
-            Debug.Assert(basesBeingResolved == null || !basesBeingResolved.ContainsReference(this.OriginalDefinition));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs:233]: >Debug.Assert(basesBeingResolved == null || !basesBeingResolved.ContainsReference(this.OriginalDefinition)); [](http://example.com/codeflow?start=12&length=107)

Why is this assert no longer valid and why it is Ok to continue if the assert fails?

**Comment by AlekseyTs** [comment]
> explicit extension B for C : A<B.Nested> { }

I am curious how does this work. It looks like we need to construct A before we can look in it.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:9593 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> explicit extension B for C : A<B.Nested> { }

Can we assert what is this symbol at the end? Basically assert what is the base extension for ```B```.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:9593 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> explicit extension B for C : A<B.Nested> { }

Please assert what is the base extension for `B`.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:9578 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> E.M4();

It doesn't look like we are verifying IL or SemanticModel behavior for this code.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:21279 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> E.M();

Is it not interesting to try other methods in this context?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:21423 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     [MyAttribute] // 1, 2

Could you remind why we are not supposed to find this type in extension?

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:23414 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         // PROTOTYPE should we also look at members from implicit extension types?

I think we will not be able to emit references to members that are defined in extensions. 

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:23489 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> }

It doesn't look like we are testing lookup in bases of the extended type.

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:23691 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by jcouv** [comment]
> explicit extension B for C<B.Nested> : A<int> { }

Same here

---
In reply to: [1936416814](https://github.com/dotnet/roslyn/pull/71779#issuecomment-1936416814) [](http://example.com/codeflow?ancestors=1936416814)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:9608 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7797]: I'll move up and clarify. The logic for finding implicit extension type members asserts that we're not looking at an extension type. But we could hit it with `Extension.Missing`...
We disallow defining extensions and extension methods for extension types, so we can skip all this when we're looking at an extension type.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1477]: I don't think the difference is observable. I picked `ConsiderEverything` to match `ConsiderEverything2` that we use in rest of `MergeHidingLookupResults` (inside `IsDerivedType`).

**Comment by jcouv** [comment]
> }

Thanks. I'd put them in the wrong local branch

---
In reply to: [1936505728](https://github.com/dotnet/roslyn/pull/71779#issuecomment-1936505728) [](http://example.com/codeflow?ancestors=1936505728)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:23691 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs [hunk 6]
```diff
@@ -231,6 +231,7 @@ private ExtensionInfo MakeDeclaredExtensionInfo(ConsList<TypeSymbol>? basesBeing
             SourceLocation? underlyingTypeLocation = null;
 
             Debug.Assert(basesBeingResolved == null || !basesBeingResolved.ContainsReference(this.OriginalDefinition));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs:233]: >Debug [](http://example.com/codeflow?start=12&length=5)

It doesn't look like there are any meaningful changes in this file. Consider reverting to its original state.

**Comment by AlekseyTs** [comment]
> }

> Thanks. I'd put them in the wrong local branch

Still do not see tests with base interfaces

---
In reply to: [1939356899](https://github.com/dotnet/roslyn/pull/71779#issuecomment-1939356899) [](http://example.com/codeflow?ancestors=1939356899,1936505728)

---
Refers to: src/Compilers/CSharp/Test/Emit2/Semantics/ExtensionTypeTests.cs:23691 in dfa977b. [](commit_id = dfa977b91714731af4ddb6037055ce789f5c47bb, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by jcouv** [comment]
@jjonescz for review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -7429,6 +7429,12 @@ private BoundExpression ResolveToExtensionMemberIfPossible(BoundExpression expr,
         }
 
         private void LookupInstanceMember(LookupResult lookupResult, TypeSymbol leftType, bool leftIsBaseReference, string rightName, int rightArity, bool invoked, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            LookupOptions options = MakeOptionsForInstanceMemberLookup(invoked, leftIsBaseReference);
+            this.LookupMembersWithFallback(lookupResult, leftType, rightName, rightArity, ref useSiteInfo, basesBeingResolved: null, options: options);
+        }
+
+        private static LookupOptions MakeOptionsForInstanceMemberLookup(bool invoked, bool leftIsBaseReference)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7437]: I didn't see any other usage of this newly extracted method.

**Review by jjonescz — APPROVED** [review]
